src/meili/match_route.cc

```diff
@@ -208,69 +216,145 @@ bool MergeRoute(std::vector<EdgeSegment>& route, const State& source, const Stat
+    // otherwise, we use the segment's source or target because if it is a node snap, the
+    // match result can only hold one candidate, we need either side of the node.
+    new_segments[old_size].source =
+        prev_match.HasState() ? first_segment->source : prev_match.distance_along;
+    new_segments.back().last_match_idx = curr_idx;
+    new_segments.back().target =
+        curr_match.HasState() ? last_segment->target : curr_match.distance_along;
+
+    first_segment = last_segment;
+    prev_idx = curr_idx;
+  }
 }
 
-template <typename match_iterator_t>
-std::vector<EdgeSegment>
-ConstructRoute(const MapMatcher& mapmatcher, match_iterator_t begin, match_iterator_t end) {
-  if (begin == end) {
+std::vector<EdgeSegment> ConstructRoute(const MapMatcher& mapmatcher,
+                                        const std::vector<MatchResult>& match_results,
+                                        baldr::GraphReader& reader) {
+  if (match_results.empty()) {
     return {};
   }
 
   std::vector<EdgeSegment> route;
   const baldr::GraphTile* tile = nullptr;
 
   // Merge segments into route
-  for (auto prev_match = end, match = begin; match != end; match++) {
-    if (!match->HasState()) {
+  // std::deque<int> match_indices;
+  std::vector<EdgeSegment> segments;
+  std::vector<EdgeSegment> new_segments;
+  const MatchResult* prev_match{nullptr};
+  int prev_idx = -1;
+  for (int curr_idx = 0, n = static_cast<int>(match_results.size()); curr_idx < n; ++curr_idx) {
+    const MatchResult& match = match_results[curr_idx];
+
+    if (!match.edgeid.Is_Valid() || !match.HasState()) {
       continue;
     }
 
-    if (prev_match != end) {
+    if (prev_match && prev_match->HasState()) {
       const auto &prev_state = mapmatcher.state_container().state(prev_match->stateid),
-                 state = mapmatcher.state_container().state(match->stateid);
+                 state = mapmatcher.state_container().state(match.stateid);
 
       // get the route between the two states by walking edge labels backwards
       // then reverse merge the segments together which are on the same edge so we have a
       // minimum number of segments. in this case we could at minimum end up with 1 segment
-      std::vector<EdgeSegment> segments;
-      MergeRoute(segments, prev_state, state);
+      segments.clear();
+      if (!MergeRoute(prev_state, state, segments) && !route.empty()) {
+        route.back().discontinuity = true;
+        prev_idx = curr_idx;
+        prev_match = &match;
+        continue;
+      }
 
-      // TODO remove: the code is pretty mature we dont need this check its wasted cpu
-      if (!ValidateRoute(mapmatcher.graphreader(), segments.begin(), segments.end(), tile)) {
-        throw std::runtime_error("Found invalid route");
+      // trivial routes (both states are the same point) will have no path
+      // this seems to be an optimization from routing.cc
+      if (segments.empty())
```

Comment:
this was a bug. merge route above can get back a route between prev_state and current state with zero edges in it. this happens when both the state candidates are the same exact snap point (and potentially the begin and end node of two different edges but are the same actual node). this is a more common occurance with map matching using the shape you got from a route computation. the fix is simple here, since there were no edges we dont need to do anything, the previous portion of the path or next portion will include edges